{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyNjwyuGxH3/bjP+itRMzYrH"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"rwWGun3RN3PM","executionInfo":{"status":"ok","timestamp":1719772651596,"user_tz":-330,"elapsed":10163,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"colab":{"base_uri":"https://localhost:8080/"},"outputId":"83e3be8b-d8ba-4518-cb68-1bbfafd064e5"},"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting biopython\n","  Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m11.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.25.2)\n","Installing collected packages: biopython\n","Successfully installed biopython-1.84\n"]}],"source":["!pip install biopython"]},{"cell_type":"code","source":["from Bio.Seq import Seq\n","\n","sequence = Seq(\"ATTGCTAGCT\")"],"metadata":{"id":"hT4xtpUsOBkr"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(sequence)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"duHFjvehOBoN","executionInfo":{"status":"ok","timestamp":1719772651597,"user_tz":-330,"elapsed":7,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"5db177bf-1915-4941-e93f-bd71fb3b3bbd"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["ATTGCTAGCT\n"]}]},{"cell_type":"code","source":["print(\"Original Sequence:      \",sequence, \"\\n\", \"Complementary sequence:\",sequence.complement() )\n","print(\"Reverse Complementary : \",sequence.reverse_complement())"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"csEUIMWdOBru","executionInfo":{"status":"ok","timestamp":1719772651597,"user_tz":-330,"elapsed":6,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"8c8b45f2-fa11-4b96-a306-37a6a311ffdf"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Original Sequence:       ATTGCTAGCT \n"," Complementary sequence: TAACGATCGA\n","Reverse Complementary :  AGCTAGCAAT\n"]}]},{"cell_type":"markdown","source":["# 4. Sequence Method"],"metadata":{"id":"4FX-R3HzQd24"}},{"cell_type":"code","source":["for index, letter in enumerate(sequence) :\n","  print(\"{} {}\".format(index, letter))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"P19enj-VQhTM","executionInfo":{"status":"ok","timestamp":1719772651597,"user_tz":-330,"elapsed":5,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"290f4bc1-f8dd-46be-d072-17f5268c4317"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["0 A\n","1 T\n","2 T\n","3 G\n","4 C\n","5 T\n","6 A\n","7 G\n","8 C\n","9 T\n"]}]},{"cell_type":"code","source":["print(\"Length of the sequence :\",len(sequence))\n","print(\"Index of the first A :\",sequence.find(\"A\"))\n","print(\"at Index 5 the nucleotide present is :\",sequence[5])\n","print(\"Count of the letter G in the sequence :\",sequence.count(\"G\")) # Non-Overlapping\n","print(\"Count method does not considered Overlapping\" )"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"LQZqLmC9SKJ2","executionInfo":{"status":"ok","timestamp":1719772651597,"user_tz":-330,"elapsed":4,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"0a5bbb85-6a78-477c-fda8-7527c37e927d"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Length of the sequence : 10\n","Index of the first A : 0\n","at Index 5 the nucleotide present is : T\n","Count of the letter G in the sequence : 2\n","Count method does not considered Overlapping\n"]}]},{"cell_type":"markdown","source":["## % of Nucleotide in the Sequence"],"metadata":{"id":"xwRkkxJ-VJKs"}},{"cell_type":"code","source":["sequence = \"AGCTGCATGACTCCGCTGATGCGTAACGTCACCTCTCAGT\"\n","\n","# What is the percentage of Guanine and Cytosine in the seqeunce ?\n","print(\"GC Percentage :\",100*(sequence.count(\"G\") + sequence.count(\"C\"))/len(sequence))\n","\n","# What is the percentage of Adenine and thymine in the seqeunce ?\n","print(\"AT Percentage :\",100*(sequence.count(\"A\") + sequence.count(\"T\"))/len(sequence))\n","\n","# what is percentage of Adenine\n","print(\"Adenine Percentage :\",100*sequence.count(\"A\")/len(sequence))\n","# what is percentage of Thimine\n","print(\"Thymine Percentage :\",100*sequence.count(\"T\") /len(sequence))\n","# what is percentage of Guanine\n","print(\"Guanine Percentage :\",100*sequence.count(\"G\") /len(sequence))\n","# what is percentage of Cytosine\n","print(\"Cytosine Percentage :\",100*sequence.count(\"C\") /len(sequence))\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"djlnsWmbVdKF","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":1127,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"6f995877-08f3-49b5-d830-319322bf93df"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["GC Percentage : 55.0\n","AT Percentage : 45.0\n","Adenine Percentage : 20.0\n","Thymine Percentage : 25.0\n","Guanine Percentage : 22.5\n","Cytosine Percentage : 32.5\n"]}]},{"cell_type":"markdown","source":["## Build in Functions to find GC %"],"metadata":{"id":"UzLxxPqcYuiJ"}},{"cell_type":"code","source":["from Bio.SeqUtils import gc_fraction\n","print(\"GC Percentage\",100*gc_fraction(sequence))\n","print(\"AT Percentage\",100*(1-gc_fraction(sequence)))   # 1-gc_fraction(sequence)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"DqYgdr8BVdGo","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":29,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"02f5f74d-9486-4587-9348-a85263f91278"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["GC Percentage 55.00000000000001\n","AT Percentage 44.99999999999999\n"]}]},{"cell_type":"markdown","source":["# 5. Slicing Sequence and FASTA format"],"metadata":{"id":"Ix7GQd47b7Yn"}},{"cell_type":"code","source":["slice_1 = sequence[5:10]\n","ending_slice = sequence[15:]\n","starting_slice = sequence[:10]\n","print(sequence)\n","print(\"slice :\",slice_1)\n","print(\"Ending slice :\",ending_slice)\n","print(\"Starting slice :\",starting_slice)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"f5ua454IVdEZ","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":27,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"3d81364b-b6a5-42d1-cf54-3e83ff6a562f"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["AGCTGCATGACTCCGCTGATGCGTAACGTCACCTCTCAGT\n","slice : CATGA\n","Ending slice : CTGATGCGTAACGTCACCTCTCAGT\n","Starting slice : AGCTGCATGA\n"]}]},{"cell_type":"markdown","source":["## FASTA"],"metadata":{"id":"O3dMKcdwwPp2"}},{"cell_type":"code","source":["fasta = f\">name\\n{sequence}\\n\"\n","print(fasta)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"BZZkmYaSVdBy","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":26,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"8744616a-3de3-4e35-8e2d-50fd922caef2"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":[">name\n","AGCTGCATGACTCCGCTGATGCGTAACGTCACCTCTCAGT\n","\n"]}]},{"cell_type":"markdown","source":["# 6. Concatenation and Case Transformation\n"],"metadata":{"id":"2o2iDuKaw1Oe"}},{"cell_type":"code","source":["list_sequence = [Seq(\"ATG\"), Seq(\"TAG\"), Seq(\"TAA\")]\n","result = Seq(\"\")\n","\n","for s in list_sequence :\n","  result += s\n","print(result)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"d55rXopauLfM","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":25,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"16f75f9f-eb80-4ffa-9b07-a18d11cec18f"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["ATGTAGTAA\n"]}]},{"cell_type":"code","source":["# Delimiter\n","contig = [Seq(\"ATG\"), Seq(\"TAG\"), Seq(\"TAA\")]\n","delimiter = Seq(\"N\"*5)\n","\n","delimiter_join = delimiter.join(contig)\n","print(delimiter_join)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"x_NH_70RuLYD","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":24,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"76c37b6a-9319-4476-8c68-d5323330c56f"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["ATGNNNNNTAGNNNNNTAA\n"]}]},{"cell_type":"markdown","source":["## Case Transformation"],"metadata":{"id":"2MhjKjljzs0f"}},{"cell_type":"code","source":["dna_seq = Seq(\"ATGCgtca\")\n","print(\"original Sequence : \",dna_seq)\n","print(dna_seq.upper())\n","print(dna_seq.lower())\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"HWhGRSx3zr4_","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":23,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"b30a6a42-f5a6-42a0-faad-0ed1984e2c6c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["original Sequence :  ATGCgtca\n","ATGCGTCA\n","atgcgtca\n"]}]},{"cell_type":"code","source":["print(\"atg\" in dna_seq)\n","print(\"atg\" in dna_seq. lower())"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"x3ArXG0jzr2h","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":22,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"856083fd-134c-4876-a6c8-12461f4ac816"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["False\n","True\n"]}]},{"cell_type":"markdown","source":["# 7. Transcription"],"metadata":{"id":"pPQ97nIP1AQY"}},{"cell_type":"code","source":["coding_strand = Seq(\"ATGCATGCATCGATCGACGCA\")\n","print(\"coding strand:\",coding_strand)\n","template_strand = coding_strand.reverse_complement()\n","print(\"template strand:\", template_strand)\n","\n","# Transcription\n","mRNA_sequence = coding_strand.transcribe()\n","print(\"mRNA sequence:\", mRNA_sequence)\n","\n","# Reverse Transcription\n","mRNA = mRNA_sequence.back_transcribe()\n","print(\"Reverse Transcription:\", mRNA)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"FT-i4UykzrzZ","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":21,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"f2d8b1d5-6bc4-4289-838d-be3b1250db20"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["coding strand: ATGCATGCATCGATCGACGCA\n","template strand: TGCGTCGATCGATGCATGCAT\n","mRNA sequence: AUGCAUGCAUCGAUCGACGCA\n","Reverse Transcription: ATGCATGCATCGATCGACGCA\n"]}]},{"cell_type":"markdown","source":["# 8. Translation"],"metadata":{"id":"qps3m7fR5S-9"}},{"cell_type":"markdown","source":["The translation tables available in Biopython are based on those from the NCBI: https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi\n","\n","\n","By default, translation will use the standard genetic code. Suppose we are dealing with a mitochondrial sequence. We need to tell the translation function to use the relevant genetic code instead:"],"metadata":{"id":"QzHYEmN6Oqls"}},{"cell_type":"code","source":["mRNA_sequence\n","print(\"Protein Sequence : \", mRNA_sequence.translate())   # \"*\" represents stop codon"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"QgK7WdaWzrwr","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":20,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"2f023de2-b415-49fe-8739-fd55fbafcb5f"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Protein Sequence :  MHASIDA\n"]}]},{"cell_type":"code","source":["gene = Seq(\n","     \"GTGAAAAAGATGCAATCTATCGTACTCGCACTTTCCCTGGTTCTGGTCGCTCCCATGGCA\"\n","     \"GCACAGGCTGCGGAAATTACGTTAGTCCCGTCAGTAAAATTACAGATAGGCGATCGTGAT\"\n","     \"AATCGTGGCTATTACTGGGATGGAGGTCACTGGCGCGACCACGGCTGGTGGAAACAACAT\"\n","     \"TATGAATGGCGAGGCAATCGCTGGCACCTACACGGACCGCCGCCACCGCCGCGCCACCAT\"\n","     \"AAGAAAGCTCCTCATGATCATCACGGCGGTCATGGTCCAGGCAAACATCACCGCTAA\")"],"metadata":{"id":"Z9J0kmt-zrt3"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(gene.translate(table='Bacterial'))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"NwMHAbqOTX7g","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":19,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"bdc59be5-73f4-494a-eb7a-c3a70a02cc8c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["VKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYWDGGHWRDHGWWKQHYEWRGNRWHLHGPPPPPRHHKKAPHDHHGGHGPGKHHR*\n"]}]},{"cell_type":"code","source":[" print(gene.translate(table='Bacterial',cds = True))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"euGS6j61O4a8","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":18,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"3a47485e-eae4-4646-9ba6-7019f158404c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["MKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYWDGGHWRDHGWWKQHYEWRGNRWHLHGPPPPPRHHKKAPHDHHGGHGPGKHHR\n"]}]},{"cell_type":"markdown","source":["# 9. Mutable Sequence"],"metadata":{"id":"kIzcw-y5PKfu"}},{"cell_type":"code","source":["from Bio.Seq import MutableSeq\n","seq = Seq(\"ACGATAGTCATGGGCCGCTGAAAGGGTGCCCGA\")\n","seq"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"rNn5cQmfO4TT","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":17,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"053d46f2-458f-4855-9ab8-bb8cb63071f1"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["Seq('ACGATAGTCATGGGCCGCTGAAAGGGTGCCCGA')"]},"metadata":{},"execution_count":21}]},{"cell_type":"code","source":["mutable_seq = MutableSeq(seq)\n","mutable_seq"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"k86b0REOO337","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":16,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"bfd1e3c0-b610-4bb3-ce1b-a65e093f80d7"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["MutableSeq('ACGATAGTCATGGGCCGCTGAAAGGGTGCCCGA')"]},"metadata":{},"execution_count":22}]},{"cell_type":"code","source":["mutable_seq[1] = \"T\"\n","mutable_seq"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ZlP7aJzOO3MS","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":16,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"38c19fc5-0cdc-4bd4-cadb-682a9f98f949"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["MutableSeq('ATGATAGTCATGGGCCGCTGAAAGGGTGCCCGA')"]},"metadata":{},"execution_count":23}]},{"cell_type":"code","source":["mutable_seq.reverse()"],"metadata":{"id":"f_JKOsJrSN3U"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["mutable_seq"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"5snlbeSASTej","executionInfo":{"status":"ok","timestamp":1719772652721,"user_tz":-330,"elapsed":14,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"fcdd8129-42ca-4215-ce63-815cdb076b75"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["MutableSeq('AGCCCGTGGGAAAGTCGCCGGGTACTGATAGTA')"]},"metadata":{},"execution_count":25}]},{"cell_type":"markdown","source":["# 10 Introduction to Sequence Record"],"metadata":{"id":"jGX0fB-mCLku"}},{"cell_type":"code","source":["from Bio.SeqRecord import SeqRecord\n","\n","# 1) Example letter_annotations\n","dna_sequence = Seq(\"ATGCTAGCTATAG\")\n","# Create a SeqRecord object\n","seq_record = SeqRecord(dna_sequence)\n","# Quality scores for each base\n","quality_scores = [30, 40, 25, 35, 20, 38, 42, 27, 39, 45, 30, 28, 36]\n","# Assign quality scores to the letter_annotations attribute\n","seq_record.letter_annotations[\"phred_quality\"] = quality_scores\n","\n","# 2) Example annotations\n","# Assign annotations (information) for the entire sequence\n","seq_record.annotations[\"organism\"] = \"Homo sapiens\"\n","seq_record.annotations[\"source\"] = \"GenBank\"\n","\n","# 3) Example features\n","from Bio.SeqFeature import FeatureLocation, SeqFeature\n","# Adding a Feature\n","feature_location = FeatureLocation(start=1, end=4, strand =1)\n","feature = SeqFeature(location = feature_location, type=\"gene\", id =\"gene123\")\n","seq_record.features.append(feature)\n","\n","# 1 or +: Indicates the positive strand (5' to 3')\n","# -1 or -: Indicates the negative strand (3' to 5')\n","\n","# 4) Example dbxrefs\n","# Adding database cross-references\n","seq_record.dbxrefs.append(\"GenBank:AC123456\")\n","seq_record.dbxrefs.append(\"EMBL:123456\")\n","\n","# 5) Adding ID\n","seq_record.id = \"AC123456\"\n","\n","# 6) adding name\n","seq_record.name = \"Human gene\"\n","\n","# 7) Adding Description\n","seq_record.description = \"Example of a gene fragment\""],"metadata":{"id":"73KdWS53SVD4"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(seq_record)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"kZ4zGjI-EaSV","executionInfo":{"status":"ok","timestamp":1719772652722,"user_tz":-330,"elapsed":14,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"211a0bf2-a673-4712-948c-a21553b8d14e"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["ID: AC123456\n","Name: Human gene\n","Description: Example of a gene fragment\n","Database cross-references: GenBank:AC123456, EMBL:123456\n","Number of features: 1\n","/organism=Homo sapiens\n","/source=GenBank\n","Per letter annotation for: phred_quality\n","Seq('ATGCTAGCTATAG')\n"]}]},{"cell_type":"markdown","source":["## Sequence Record from FASTA File\n","In this example we use a fairly large FASTA file containing the whole sequence for Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, originally downloaded from the NCBI. This file is included in the downloadable documents. The name of the file is NC_005816.fna"],"metadata":{"id":"qSuuMHv5KmzD"}},{"cell_type":"code","source":["# from Bio import SeqIO\n","# sequence_record = SeqIO.read(\"NC_005816.fna\",\"fasta\")\n","# print(sequence_record)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":321},"id":"7e-wNa17Dw0e","executionInfo":{"status":"error","timestamp":1719772652722,"user_tz":-330,"elapsed":12,"user":{"displayName":"Gaurav Pawar","userId":"05347710136466900291"}},"outputId":"eb54a971-f92e-4a12-ce1c-2f86ac10edfd"},"execution_count":null,"outputs":[{"output_type":"error","ename":"FileNotFoundError","evalue":"[Errno 2] No such file or directory: 'NC_005816.fna'","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-28-420eefa89dbe>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0msequence_record\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"NC_005816.fna\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msequence_record\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/Bio/SeqIO/__init__.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(handle, format, alphabet)\u001b[0m\n\u001b[1;32m    670\u001b[0m     \u001b[0mto\u001b[0m \u001b[0mread\u001b[0m \u001b[0mmultiple\u001b[0m \u001b[0mrecords\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mhandle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    671\u001b[0m     \"\"\"\n\u001b[0;32m--> 672\u001b[0;31m     \u001b[0miterator\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malphabet\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    673\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    674\u001b[0m         \u001b[0mrecord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterator\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/Bio/SeqIO/__init__.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(handle, format, alphabet)\u001b[0m\n\u001b[1;32m    623\u001b[0m     \u001b[0miterator_generator\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_FormatToIterator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    624\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0miterator_generator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 625\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0miterator_generator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    626\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mformat\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mAlignIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_FormatToIterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    627\u001b[0m         \u001b[0;31m# Use Bio.AlignIO to read in the alignments\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/Bio/SeqIO/FastaIO.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, source, alphabet)\u001b[0m\n\u001b[1;32m    185\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0malphabet\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"The alphabet argument is no longer supported\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"t\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfmt\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"Fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhandle\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/Bio/SeqIO/Interfaces.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, source, alphabet, mode, fmt)\u001b[0m\n\u001b[1;32m     62\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"The alphabet argument is no longer supported\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_PathLikeTypes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 64\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstream\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"r\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     65\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshould_close_stream\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     66\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'NC_005816.fna'"]}]},{"cell_type":"markdown","source":["# 13 Introduction to Sequence Feature"],"metadata":{"id":"YKm61JXBOPqh"}},{"cell_type":"markdown","source":["### Fuzzy Positions"],"metadata":{"id":"S9w9CzYPOV5C"}},{"cell_type":"code","source":["from Bio import SeqFeature\n","start_pos = SeqFeature.AfterPosition(5)\n","end_pos = SeqFeature.BetweenPosition(9,left=8, right=9)\n","\n","location = SeqFeature.SimpleLocation(start_pos, end_pos)\n","print(location)"],"metadata":{"id":"DjSUfPNmNcZc"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We can access the fuzzy start and end positions"],"metadata":{"id":"_pRqYfflPF0Z"}},{"cell_type":"code","source":["print(location.start)\n","print(location.end)"],"metadata":{"id":"07fndwDRNcWp"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["### Simple Positions"],"metadata":{"id":"cCp_FxiAPVBQ"}},{"cell_type":"code","source":["exact_location = SeqFeature.SimpleLocation(5, 12)\n","print(exact_location)"],"metadata":{"id":"DjYo8jaCNcT-"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["exact_location.start\n"],"metadata":{"id":"xPPkIuagNcQ7"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["exact_location.end"],"metadata":{"id":"307D3AFwNcNb"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# 14 SNP detection"],"metadata":{"id":"mhZBPvuNP_fc"}},{"cell_type":"code","source":["snp_index = 4352\n","record = SeqIO.read(\"NC_005816.gb\",\"genbank\")\n","for feature in record.features :\n","  if snp_index in feature :\n","    print(\"{} {}\".format(feature.type , feature.qualifiers.get(\"db_xref\")))"],"metadata":{"id":"7NwCDvAHDwuv"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# 15 Comparison"],"metadata":{"id":"HMr-yR1ySjb9"}},{"cell_type":"code","source":["record1  = SeqRecord(Seq(\"ACGT\"), id = \"record1\")\n","record2 = SeqRecord(Seq(\"ACGT\"), id = \"record2\")\n","\n","print(record1.id == record2.id)\n","print(record1.seq == record2.seq)"],"metadata":{"id":"Lk-FS87EP22Q"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# 16 Format Method"],"metadata":{"id":"tmew_BUPTPZc"}},{"cell_type":"code","source":["sequence_data = \"MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD\" \\\n","                 \"GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK\" \\\n","                 \"NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM\" \\\n","                 \"SSAC\"\n","record = SeqRecord(Seq(sequence_data),\n","                   id = \"gi|14150838|gb|AAK54648.1|AF376133_1\",\n","                   description = \"chalcone synthase [Cucumis sativus]\"\n","                   )\n","print(record.format(\"fasta\"))\n","print(record)"],"metadata":{"id":"qBLdkOf2P2xG"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# 17 Slicing a Sequence Record"],"metadata":{"id":"Lb1oeQfPTgJ4"}},{"cell_type":"markdown","source":["**Slicing** is a concept in programming and data manipulation that involves extracting a portion or subset of data from a larger sequence. We are going to use the *NC_005816.gb* as an example."],"metadata":{"id":"bk3KNdg2Tx9L"}},{"cell_type":"code","source":["record =SeqIO.read(\"NC_005816.gb\",\"genbank\")\n","print(\"length of the sequnece :\",len(record))\n","print(f\"length of Record feature : {len(record.features)}\")"],"metadata":{"id":"srFR6iD0ToID"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We are going to focus in on the pim gene, YP_pPCP05"],"metadata":{"id":"HCsNXPY-UBfz"}},{"cell_type":"code","source":["print(record.features[20])"],"metadata":{"id":"VzgQe0X5TnAQ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(record.features[21])"],"metadata":{"id":"y6s9gZ39TqIG"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["record.annotations[\"topology\"]"],"metadata":{"id":"N9KRhtjLW90T"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Let us **slice** this parent record from 4300 to 4800 (enough to include the pim gene/CDS).\n","\n","Let us analyze how many features we get."],"metadata":{"id":"Vhas-ocmXZYC"}},{"cell_type":"code","source":["sliced = record[4300:4800]\n","print(sliced)\n","print(f\"length of sliced sequence : {len(sliced)}\")"],"metadata":{"id":"M6t8MCzoW9zK"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["sliced.annotations[\"topology\"] = \"linear\""],"metadata":{"id":"LigcktjCX1DC"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(sliced.description)\n","sliced.description = \"Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, partial sequence\"\n","print(sliced.description)"],"metadata":{"id":"F3DwEZkPX1Ap"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# 18 Adding a sequence record"],"metadata":{"id":"cDqQ4--4YZaM"}},{"cell_type":"code","source":["record = next(SeqIO.parse(\"example.fastq\",\"fastq\"))\n","print(\"length :\",len(record))\n","print(record.seq)\n","print(record.letter_annotations[\"phred_quality\"])"],"metadata":{"id":"I70McOQiX09q"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Let us suppose this was Roche 454 data, and that from other source we think the **TTT** should be only **TT**."],"metadata":{"id":"XEcLMTSNZh2Q"}},{"cell_type":"code","source":["left = record[:19]\n","print(left.seq)\n","print(left.letter_annotations[\"phred_quality\"])"],"metadata":{"id":"yQoA9fVNZRGr"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["right = record[20:]\n","print(right.seq)\n","print(right.letter_annotations[\"phred_quality\"])"],"metadata":{"id":"9M0dyAOVX07G"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We now **add** the two parts together"],"metadata":{"id":"hDvMgZHIZ23n"}},{"cell_type":"code","source":["updated_seq = left +right\n","print(len(updated_seq))\n","print(updated_seq)\n","print(updated_seq.letter_annotations[\"phred_quality\"])\n","\n","# or we can write\n","# updated_seq = record[:19] + record[20:]"],"metadata":{"id":"hOh4oOpnX04Y"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# 19 Reverse Compliment sequqnce"],"metadata":{"id":"2SZHubleaZvK"}},{"cell_type":"code","source":["# We are going to use the genbank file **NC_005816.gb**\n","record = SeqIO.read(\"NC_005816.gb\",\"genbank\")\n","print(f\"id : {record.id}\\n length : {len(record)}\\n Feature : {len(record.features)}\\\n","  \\n Reference : {len(record.dbxrefs)} \\n Annotations :{len(record.annotations)}\\n\",\"---\"*10)\n","\n","#  let us reverse complement this sequence record\n","rc = record.reverse_complement(id = \"Reverse_Complement\")\n","print(rc)"],"metadata":{"id":"OMdLVLX2X01y"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# 20 Parsing and Reading Sequences"],"metadata":{"id":"cbfws1a5cemq"}},{"cell_type":"code","source":["# We are going to work with two examples: ls_orchid.fasta and ls_orchid.gbk\n","# for sequence_record in SeqIO.parse(\"ls_orchid.gbk\",\"genbank\") :\n","for sequence_record in SeqIO.parse(\"ls_orchid.fasta\",\"fasta\") :\n","  print(sequence_record.id)\n","  print(repr(sequence_record.seq))\n","  print(len(seq_record))"],"metadata":{"collapsed":true,"id":"dpFAvWrvW9xA"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["identifiers  = [sequence_record. id for sequence_record in SeqIO.parse(\"ls_orchid.fasta\",\"fasta\")]\n","identifiers"],"metadata":{"collapsed":true,"id":"JASSYQineC20"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# 21 Iterating over a sequence file\n"],"metadata":{"id":"J6WLu1F1eiWE"}},{"cell_type":"markdown","source":["We can use the **next()** function on an iterator to step through the entries"],"metadata":{"id":"b1h3jk60esHB"}},{"cell_type":"code","source":["record_iterator = SeqIO.parse(\"ls_orchid.fasta\",\"fasta\")"],"metadata":{"id":"2nfs_eaIetz3"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["first_iterator = next(record_iterator)\n","print(first_iterator)"],"metadata":{"id":"9yXGg8TqetxM"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["second_iterator = next(record_iterator)\n","print(second_iterator)"],"metadata":{"id":"ot7mIVMjW9wG"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# 23 Writing a Sequence File"],"metadata":{"id":"0dF0CpnUsD2L"}},{"cell_type":"markdown","source":["### Writing to a File Example"],"metadata":{"id":"ttyKl_c6tO75"}},{"cell_type":"markdown","source":[],"metadata":{"id":"zIvnkbuPtsaJ"}},{"cell_type":"code","source":["rec1 = SeqRecord(\n","    Seq(\n","        \"MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD\"\n","        \"GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK\"\n","        \"NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM\"\n","        \"SSAC\",\n","    ),\n","    id=\"gi|14150838|gb|AAK54648.1|AF376133_1\",\n","    description=\"chalcone synthase [Cucumis sativus]\",\n",")\n","\n","rec2 = SeqRecord(\n","    Seq(\n","        \"YPDYYFRITNREHKAELKEKFQRMCDKSMIKKRYMYLTEEILKENPSMCEYMAPSLDARQ\"\n","        \"DMVVVEIPKLGKEAAVKAIKEWGQ\",\n","    ),\n","    id=\"gi|13919613|gb|AAK33142.1|\",\n","    description=\"chalcone synthase [Fragaria vesca subsp. bracteata]\",\n",")\n","\n","rec3 = SeqRecord(\n","    Seq(\n","        \"MVTVEEFRRAQCAEGPATVMAIGTATPSNCVDQSTYPDYYFRITNSEHKVELKEKFKRMC\"\n","        \"EKSMIKKRYMHLTEEILKENPNICAYMAPSLDARQDIVVVEVPKLGKEAAQKAIKEWGQP\"\n","        \"KSKITHLVFCTTSGVDMPGCDYQLTKLLGLRPSVKRFMMYQQGCFAGGTVLRMAKDLAEN\"\n","        \"NKGARVLVVCSEITAVTFRGPNDTHLDSLVGQALFGDGAAAVIIGSDPIPEVERPLFELV\"\n","        \"SAAQTLLPDSEGAIDGHLREVGLTFHLLKDVPGLISKNIEKSLVEAFQPLGISDWNSLFW\"\n","        \"IAHPGGPAILDQVELKLGLKQEKLKATRKVLSNYGNMSSACVLFILDEMRKASAKEGLGT\"\n","        \"TGEGLEWGVLFGFGPGLTVETVVLHSVAT\",\n","    ),\n","    id=\"gi|13925890|gb|AAK49457.1|\",\n","    description=\"chalcone synthase [Nicotiana tabacum]\",\n",")"],"metadata":{"id":"T61jTPGyW9r9"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["my_records = [rec1,rec2,rec3]\n","SeqIO.write(my_records,\"example.faa\",\"fasta\")"],"metadata":{"id":"WAY8RwOqW9qA"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["  # 24 Converting a File"],"metadata":{"id":"8bBLujhP3Fxn"}},{"cell_type":"code","source":["records = SeqIO.parse(\"ls_orchid.gbk\",\"genbank\")\n","count = SeqIO.write(records,\"example1.fasta\",\"fasta\")\n","print(f\"Number of Records converted : {count}\")"],"metadata":{"id":"qMTylJLPW9om"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["There is a helper function that lets us replace the previous work with:"],"metadata":{"id":"Mv76SLwx43gV"}},{"cell_type":"code","source":["count = SeqIO.convert(\"ls_orchid.gbk\",\"genbank\",\"example3.fasta\",\"fasta\")\n","print(f\"Number of Records converted : {count}\")"],"metadata":{"id":"oums4qXvW9nr"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Converting a File of Sequences to their Reverse Complements\n"],"metadata":{"id":"iE8XxNrz6naU"}},{"cell_type":"code","source":["for record in SeqIO.parse(\"ls_orchid.gbk\",\"genbank\") :\n","  print(record.id)\n","  print(\"Original Sequenece :\")\n","  print(record.seq)\n","  print(\"Reverse Complement :\")\n","  print(record.reverse_complement())"],"metadata":{"collapsed":true,"id":"gJTEdxoSW9kA"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["rev_c_records = [rec.reverse_complement(id=\"rc_\" + rec.id,description = \"rc\")\n","                for rec in SeqIO.parse(\"ls_orchid.fasta\",\"fasta\")]"],"metadata":{"id":"8mDw29NN6p1x"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["rev_c_records"],"metadata":{"id":"uVMrcP1N6puX","collapsed":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["len(rev_c_records)"],"metadata":{"id":"uMVY9ghB0NW4"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["rev_c_records2 = [rec.reverse_complement(id=\"rc_\" + rec.id,description = \"rc\")\n","                for rec in SeqIO.parse(\"ls_orchid.gbk\",\"genbank\")\n","                if len(rec) < 650]"],"metadata":{"id":"E6eP7Dg80NaV"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["len(rev_c_records2)"],"metadata":{"id":"5UdlZ76S6prq"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# 25 Introdction to Sequence Alignment"],"metadata":{"id":"xHl7Xi7A1MzM"}},{"cell_type":"code","source":["from Bio.Align import Alignment\n","import numpy as np"],"metadata":{"id":"LtB40v_QW9JM"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["seqA = \"CCGGTTTTT\"\n","seqB = \"AGTTTAA\"\n","seqC = \"AGGTTT\"\n","\n","sequences = [seqA, seqB, seqC]"],"metadata":{"id":"yqWkkBnHW3CS"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["coords = np.array([[1, 3, 4, 7, 9], [0, 2, 2, 5, 5], [0, 2, 3, 6, 6]])"],"metadata":{"id":"9FnR8-N91jJh"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["These coordinates define the alignment for the following sequence segments:\n","\n","*   SeqA[1:3], SeqB[0:2], and SeqC[0:2] are aligned to each other;\n","*   SeqA[3:4] and SeqC[2:3] are aligned to each other, with a gap of one nucleotide in seqB;\n","*  SeqA[4:7], SeqB[2:5], and SeqC[3:6] are aligned to each other;\n","*  SeqA[7:9] is not aligned to seqB or seqC.\n","\n"],"metadata":{"id":"haLx7u802ydw"}},{"cell_type":"markdown","source":["Let us now create the Alignment object."],"metadata":{"id":"yB-4mWn-23Gg"}},{"cell_type":"code","source":["alignment = Alignment(sequences,coords)\n","alignment"],"metadata":{"id":"31aUh31k1jG8"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The alignment object has an attribute *sequences* pointing to the sequences included in this alignment.\n"],"metadata":{"id":"flpWkVp228Yh"}},{"cell_type":"code","source":["alignment.sequences"],"metadata":{"id":"6ibT_6V7254N"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["alignment.coordinates"],"metadata":{"id":"u_BEQUa9251k"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(alignment)"],"metadata":{"id":"mtfqZBmc25zC"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Slicing and Indexing\n"],"metadata":{"id":"9UhKz4wY3RLz"}},{"cell_type":"markdown","source":["####Retrieve Entire Sequences"],"metadata":{"id":"XJuaL7NI3UgC"}},{"cell_type":"code","source":["alignment[0]"],"metadata":{"id":"bIcgUEAf3UCV"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["#### Retrieve Specific Segments:\n","`alignment[k, i:j]`: Returns a substring of the sequence from row `k`, starting at column `i` and ending at column `j-1`."],"metadata":{"id":"r24f1oLa3bus"}},{"cell_type":"code","source":["# Original first row : CGGTTTTT\n","print(alignment[0,:])\n","print(alignment[0,1:])\n","print(alignment[0,2:])\n","print()\n","print(alignment[1,:])\n","print(alignment[1,1:])\n","print(alignment[1,2:])"],"metadata":{"id":"s4eMx9bj25uf"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Retrieve Specific Columns:\n","print(alignment[:,0],\n","alignment[:,2],\n","alignment[:,3])"],"metadata":{"id":"kdV0eFWV25r8"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Create new Alignment objects from Slices:\n","a1 = alignment[1:]\n","print(a1)"],"metadata":{"id":"psBgIFoD25lT"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# 28 Getting information about alignment"],"metadata":{"id":"hg6nSk_E4oqz"}},{"cell_type":"code","source":["# Alignment Shape\n","print(len(alignment))   # the number of aligned sequence\n","print(alignment.length)  # The alignment length is defined as the number of columns in the alignment as printed\n","print(alignment.shape)  # returns a tuple consisting of number of aligned sequences and the length"],"metadata":{"id":"ByWjOkIJ25ip"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["#### Comparing Alignments\n","Two alignments are equal to each other (meaning that `alignment1 == alignment2` evaluates to *True*) if each of the sequences in `alignment1.sequences` and `alignment2.sequences` are equal to each other, and `alignment1.coordinates` and `alignment2.coordinates` contain the same coordinates. If either of these conditions is not fulfilled, then `alignment1 == alignment2` evaluates to `False`.\n","\n"],"metadata":{"id":"SjeNpqL27-Db"}},{"cell_type":"markdown","source":["#### Finding the Indices of Aligned Sequences\n","For pairwise alignments, the `aligned` attribute of an alignment returns the start and end indices of subsequences in the target and query sequence that were aligned to each other. If the alignment between target (*t*) and query (*q*) consists of *N* chunks, you get two tuples of length *N*."],"metadata":{"id":"_siWMxsp8FRq"}},{"cell_type":"code","source":["pairwiseAlignment = alignment[:2,:]\n","print(pairwiseAlignment)\n","pairwiseAlignment.aligned"],"metadata":{"id":"k1Yg2QUA8HYf"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Different alignments can result in the same subsequences being aligned with each other. This often happens when the alignments vary solely in their placement of gaps."],"metadata":{"id":"MUB6NE188hPu"}},{"cell_type":"code","source":[],"metadata":{"id":"8v5G12sx8fPj"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["from Bio import SeqIO\n","from Bio import Entrez\n","Entrez.email = \"gaurav.gnp.412@gmail.com\"\n","\n","handle = Entrez.efetch(db=\"nucleotide\", id= \"M30794.1\", rettype=\"gb\",retmode=\"text\")\n","\n","record = SeqIO.read(handle,\"genbank\")\n","\n","handle.close()"],"metadata":{"id":"pUP2bvgl8a56"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(record)"],"metadata":{"id":"OOwtM7nf8HWc"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["len(record.features)"],"metadata":{"id":"vLVe8iRC8HT0"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(record.features[6])"],"metadata":{"id":"hKCTX-Rj8HNR"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"4CXDrnYYlBYt"},"execution_count":null,"outputs":[]}]}